SplitBrainHandlerTest failures

```
java.lang.AssertionError: CountDownLatch failed to complete within 120 seconds , count left: 1
	at org.junit.Assert.fail(Assert.java:88)
	at org.junit.Assert.assertTrue(Assert.java:41)
	at com.hazelcast.test.HazelcastTestSupport.assertOpenEventually(HazelcastTestSupport.java:758)
	at com.hazelcast.test.HazelcastTestSupport.assertOpenEventually(HazelcastTestSupport.java:751)
	at com.hazelcast.cluster.SplitBrainHandlerTest.testClusterMerge_ignoresLiteMembers(SplitBrainHandlerTest.java:577)
```

https://hazelcast-l337.ci.cloudbees.com/view/Official%20Builds/job/Hazelcast-3.maintenance-nightly/com.hazelcast$hazelcast/849/testReport/junit/com.hazelcast.cluster/SplitBrainHandlerTest/testClusterMerge_ignoresLiteMembers/

**testMulticast_ClusterMerge_when_split_not_detected_by_master
testClusterMerge_ignoresLiteMembers
testClusterMerge_when_split_not_detected_by_slave_and_restart_during_merge
testClusterMerge_when_split_not_detected_by_slave**
```
java.lang.AssertionError: CountDownLatch failed to complete within 120 seconds , count left: 3
	at org.junit.Assert.fail(Assert.java:88)
	at org.junit.Assert.assertTrue(Assert.java:41)
	at com.hazelcast.test.HazelcastTestSupport.assertOpenEventually(HazelcastTestSupport.java:764)
	at com.hazelcast.test.HazelcastTestSupport.assertOpenEventually(HazelcastTestSupport.java:757)
	at com.hazelcast.cluster.SplitBrainHandlerTest.testClusterMerge_ignoresLiteMembers(SplitBrainHandlerTest.java:575)
```

**testTcpIp_ClusterMerge_when_split_not_detected_by_master**
```
java.lang.AssertionError: null
	at org.junit.Assert.fail(Assert.java:86)
	at org.junit.Assert.assertTrue(Assert.java:41)
	at org.junit.Assert.assertTrue(Assert.java:52)
	at com.hazelcast.cluster.SplitBrainHandlerTest.testClusterMerge_when_split_not_detected_by_master(SplitBrainHandlerTest.java:503)
	at com.hazelcast.cluster.SplitBrainHandlerTest.testMulticast_ClusterMerge_when_split_not_detected_by_master(SplitBrainHandlerTest.java:432)
```

https://hazelcast-l337.ci.cloudbees.com/view/Official%20Builds/job/Hazelcast-3.x-nightly/953/#showFailuresLink

**testClusterMerge_when_split_not_detected_by_slave_and_restart_during_merge**
```
java.lang.AssertionError: null
	at org.junit.Assert.fail(Assert.java:86)
	at org.junit.Assert.assertTrue(Assert.java:41)
	at org.junit.Assert.assertTrue(Assert.java:52)
	at com.hazelcast.cluster.SplitBrainHandlerTest.testClusterMerge_when_split_not_detected_by_slave_and_restart_during_merge(SplitBrainHandlerTest.java:839)
```

https://hazelcast-l337.ci.cloudbees.com/view/Official%20Builds/job/Hazelcast-3.x-nightly/com.hazelcast$hazelcast/963/testReport/junit/com.hazelcast.cluster/SplitBrainHandlerTest/testClusterMerge_when_split_not_detected_by_slave_and_restart_during_merge/

**testClusterMerge_when_split_not_detected_by_slave**

```
java.lang.AssertionError: null
	at org.junit.Assert.fail(Assert.java:86)
	at org.junit.Assert.assertTrue(Assert.java:41)
	at org.junit.Assert.assertTrue(Assert.java:52)
	at com.hazelcast.cluster.SplitBrainHandlerTest.testClusterMerge_when_split_not_detected_by_slave(SplitBrainHandlerTest.java:751)
```

https://hazelcast-l337.ci.cloudbees.com/view/Official%20Builds/job/Hazelcast-3.x-nightly/com.hazelcast$hazelcast/964/testReport/junit/com.hazelcast.cluster/SplitBrainHandlerTest/testClusterMerge_when_split_not_detected_by_slave/

**testClusterMerge_when_split_not_detected_by_slave_and_restart_during_merge**

```
java.lang.AssertionError: null
	at org.junit.Assert.fail(Assert.java:86)
	at org.junit.Assert.assertTrue(Assert.java:41)
	at org.junit.Assert.assertTrue(Assert.java:52)
	at com.hazelcast.cluster.SplitBrainHandlerTest.testClusterMerge_when_split_not_detected_by_slave_and_restart_during_merge(SplitBrainHandlerTest.java:839)
```

https://hazelcast-l337.ci.cloudbees.com/view/Official%20Builds/job/Hazelcast-3.x-nightly/

**testTcpIp_ClusterMerge_when_split_not_detected_by_master**
```
java.lang.AssertionError: null
	at org.junit.Assert.fail(Assert.java:86)
	at org.junit.Assert.assertTrue(Assert.java:41)
	at org.junit.Assert.assertTrue(Assert.java:52)
	at com.hazelcast.cluster.SplitBrainHandlerTest.testClusterMerge_when_split_not_detected_by_master(SplitBrainHandlerTest.java:503)
	at com.hazelcast.cluster.SplitBrainHandlerTest.testTcpIp_ClusterMerge_when_split_not_detected_by_master(SplitBrainHandlerTest.java:437)
```

https://hazelcast-l337.ci.cloudbees.com/view/Hazelcast/job/Hazelcast-3.x-nightly/com.hazelcast$hazelcast/968/testReport/junit/com.hazelcast.cluster/SplitBrainHandlerTest/testTcpIp_ClusterMerge_when_split_not_detected_by_master/

**testClusterMerge_when_split_not_detected_by_slave_and_restart_during_merge**
```
java.lang.AssertionError: null
	at org.junit.Assert.fail(Assert.java:86)
	at org.junit.Assert.assertTrue(Assert.java:41)
	at org.junit.Assert.assertTrue(Assert.java:52)
	at com.hazelcast.cluster.SplitBrainHandlerTest.testClusterMerge_when_split_not_detected_by_slave_and_restart_during_merge(SplitBrainHandlerTest.java:839)
```

https://hazelcast-l337.ci.cloudbees.com/view/Official%20Builds/job/Hazelcast-3.x-nightly/com.hazelcast$hazelcast/970/testReport/junit/com.hazelcast.cluster/SplitBrainHandlerTest/testClusterMerge_when_split_not_detected_by_slave_and_restart_during_merge/

**testClusterMerge_when_split_not_detected_by_slave**

```
java.lang.AssertionError: null
	at org.junit.Assert.fail(Assert.java:86)
	at org.junit.Assert.assertTrue(Assert.java:41)
	at org.junit.Assert.assertTrue(Assert.java:52)
	at com.hazelcast.cluster.SplitBrainHandlerTest.testClusterMerge_when_split_not_detected_by_slave(SplitBrainHandlerTest.java:751)
```

https://hazelcast-l337.ci.cloudbees.com/view/Official%20Builds/job/Hazelcast-3.x-nightly/com.hazelcast$hazelcast/971/testReport/junit/com.hazelcast.cluster/SplitBrainHandlerTest/testClusterMerge_when_split_not_detected_by_slave/

**testClusterMerge_when_split_not_detected_by_slave_and_restart_during_merge**
```
java.lang.AssertionError: null
	at org.junit.Assert.fail(Assert.java:86)
	at org.junit.Assert.assertTrue(Assert.java:41)
	at org.junit.Assert.assertTrue(Assert.java:52)
	at com.hazelcast.cluster.SplitBrainHandlerTest.testClusterMerge_when_split_not_detected_by_slave_and_restart_during_merge(SplitBrainHandlerTest.java:839)
```

https://hazelcast-l337.ci.cloudbees.com/view/Official%20Builds/job/Hazelcast-3.x-nightly/com.hazelcast$hazelcast/972/testReport/junit/com.hazelcast.cluster/SplitBrainHandlerTest/testClusterMerge_when_split_not_detected_by_slave_and_restart_during_merge/

**testTcpIp_ClusterMerge_when_split_not_detected_by_master**
```
java.lang.AssertionError: null
	at org.junit.Assert.fail(Assert.java:86)
	at org.junit.Assert.assertTrue(Assert.java:41)
	at org.junit.Assert.assertTrue(Assert.java:52)
	at com.hazelcast.cluster.SplitBrainHandlerTest.testClusterMerge_when_split_not_detected_by_master(SplitBrainHandlerTest.java:503)
	at com.hazelcast.cluster.SplitBrainHandlerTest.testTcpIp_ClusterMerge_when_split_not_detected_by_master(SplitBrainHandlerTest.java:437)
```

https://hazelcast-l337.ci.cloudbees.com/view/Official%20Builds/job/Hazelcast-3.x-nightly/com.hazelcast$hazelcast/976/testReport/junit/com.hazelcast.cluster/SplitBrainHandlerTest/testTcpIp_ClusterMerge_when_split_not_detected_by_master/

```
java.lang.AssertionError: null
	at org.junit.Assert.fail(Assert.java:86)
	at org.junit.Assert.assertTrue(Assert.java:41)
	at org.junit.Assert.assertTrue(Assert.java:52)
	at com.hazelcast.cluster.SplitBrainHandlerTest.testClusterMerge_when_split_not_detected_by_slave(SplitBrainHandlerTest.java:751)
```

https://hazelcast-l337.ci.cloudbees.com/view/Official%20Builds/job/Hazelcast-3.x-nightly/com.hazelcast$hazelcast/979/testReport/junit/com.hazelcast.cluster/SplitBrainHandlerTest/testClusterMerge_when_split_not_detected_by_slave/

**testMulticast_ClusterMerge_when_split_not_detected_by_master**
```
java.lang.AssertionError: null
	at org.junit.Assert.fail(Assert.java:86)
	at org.junit.Assert.assertTrue(Assert.java:41)
	at org.junit.Assert.assertTrue(Assert.java:52)
	at com.hazelcast.cluster.SplitBrainHandlerTest.testClusterMerge_when_split_not_detected_by_master(SplitBrainHandlerTest.java:503)
	at com.hazelcast.cluster.SplitBrainHandlerTest.testMulticast_ClusterMerge_when_split_not_detected_by_master(SplitBrainHandlerTest.java:432)
```

https://hazelcast-l337.ci.cloudbees.com/view/Official%20Builds/job/Hazelcast-3.x-nightly/com.hazelcast$hazelcast/982/testReport/junit/com.hazelcast.cluster/SplitBrainHandlerTest/testMulticast_ClusterMerge_when_split_not_detected_by_master/

--- a/hazelcast/src/main/java/com/hazelcast/internal/cluster/impl/AbstractJoiner.java
+++ b/hazelcast/src/main/java/com/hazelcast/internal/cluster/impl/AbstractJoiner.java
@@ -20,6 +20,7 @@ import com.hazelcast.cluster.ClusterState;
 import com.hazelcast.cluster.Joiner;
 import com.hazelcast.config.Config;
 import com.hazelcast.core.Member;
+import com.hazelcast.core.MemberLeftException;
 import com.hazelcast.instance.Node;
 import com.hazelcast.instance.NodeState;
 import com.hazelcast.internal.cluster.ClusterService;
@@ -35,7 +36,9 @@ import com.hazelcast.spi.OperationService;
 import com.hazelcast.spi.properties.GroupProperty;
 import com.hazelcast.util.Clock;
 import com.hazelcast.util.EmptyStatement;
+import com.hazelcast.util.FutureUtil;
 
+import java.util.ArrayList;
 import java.util.Collection;
 import java.util.concurrent.ConcurrentHashMap;
 import java.util.concurrent.ConcurrentMap;
@@ -47,6 +50,7 @@ import java.util.concurrent.atomic.AtomicLong;
 
 import static com.hazelcast.cluster.memberselector.MemberSelectors.DATA_MEMBER_SELECTOR;
 import static com.hazelcast.spi.impl.OperationResponseHandlerFactory.createEmptyResponseHandler;
+import static com.hazelcast.util.FutureUtil.waitWithDeadline;
 
 public abstract class AbstractJoiner implements Joiner {
 
@@ -55,6 +59,7 @@ public abstract class AbstractJoiner implements Joiner {
     private static final long SPLIT_BRAIN_CONN_TIMEOUT = 5000;
     private static final long SPLIT_BRAIN_SLEEP_TIME = 10;
     private static final int SPLIT_BRAIN_JOIN_CHECK_TIMEOUT_SECONDS = 10;
+    private static final int SPLIT_BRAIN_MERGE_TIMEOUT_SECONDS = 30;
 
     protected final Config config;
     protected final Node node;
@@ -70,6 +75,17 @@ public abstract class AbstractJoiner implements Joiner {
 
     private final long mergeNextRunDelayMs;
     private volatile Address targetAddress;
+
+    private final FutureUtil.ExceptionHandler splitBrainMergeExceptionHandler = new FutureUtil.ExceptionHandler() {
+        @Override
+        public void handleException(Throwable throwable) {
+            if (throwable instanceof MemberLeftException) {
+                return;
+            }
+            logger.warning("Problem while waiting for merge operation result", throwable);
+        }
+    };
+
 
     public AbstractJoiner(Node node) {
         this.node = node;
@@ -329,12 +345,17 @@ public abstract class AbstractJoiner implements Joiner {
 
         OperationService operationService = node.nodeEngine.getOperationService();
         Collection<Member> memberList = clusterService.getMembers();
+        Collection<Future> futures = new ArrayList<Future>(memberList.size());
         for (Member member : memberList) {
             if (!member.localMember()) {
                 Operation op = new MergeClustersOperation(targetAddress);
-                operationService.invokeOnTarget(ClusterServiceImpl.SERVICE_NAME, op, member.getAddress());
+                Future<Object> future =
+                        operationService.invokeOnTarget(ClusterServiceImpl.SERVICE_NAME, op, member.getAddress());
+                futures.add(future);
             }
         }
+
+        waitWithDeadline(futures, SPLIT_BRAIN_MERGE_TIMEOUT_SECONDS, TimeUnit.SECONDS, splitBrainMergeExceptionHandler);
 
         Operation mergeClustersOperation = new MergeClustersOperation(targetAddress);
         mergeClustersOperation.setNodeEngine(node.nodeEngine).setService(clusterService)

--- a/hazelcast/src/test/java/com/hazelcast/cluster/SplitBrainHandlerTest.java
+++ b/hazelcast/src/test/java/com/hazelcast/cluster/SplitBrainHandlerTest.java
@@ -44,7 +44,6 @@ import com.hazelcast.test.annotation.NightlyTest;
 import com.hazelcast.util.Clock;
 import org.junit.After;
 import org.junit.Before;
-import org.junit.Ignore;
 import org.junit.Test;
 import org.junit.experimental.categories.Category;
 import org.junit.runner.RunWith;
@@ -434,7 +433,7 @@ public class SplitBrainHandlerTest extends HazelcastTestSupport {
     }
 
     @Test
-    @Ignore //https://github.com/hazelcast/hazelcast/issues/8137
+    // https://github.com/hazelcast/hazelcast/issues/8137
     public void testTcpIp_ClusterMerge_when_split_not_detected_by_master() throws InterruptedException {
         testClusterMerge_when_split_not_detected_by_master(false);
     }
@@ -690,7 +689,7 @@ public class SplitBrainHandlerTest extends HazelcastTestSupport {
     }
 
     @Test
-    @Ignore //https://github.com/hazelcast/hazelcast/issues/8137
+    // https://github.com/hazelcast/hazelcast/issues/8137
     public void testClusterMerge_when_split_not_detected_by_slave() throws InterruptedException {
         Config config = new Config();
         String groupName = generateRandomString(10);
@@ -763,7 +762,7 @@ public class SplitBrainHandlerTest extends HazelcastTestSupport {
 
 
     @Test
-    @Ignore //https://github.com/hazelcast/hazelcast/issues/8137
+    // https://github.com/hazelcast/hazelcast/issues/8137
     public void testClusterMerge_when_split_not_detected_by_slave_and_restart_during_merge() throws InterruptedException {
         Config config = new Config();
         String groupName = generateRandomString(10);

--- a/hazelcast/src/test/java/com/hazelcast/nio/tcp/FirewallingTcpIpConnectionManager.java
+++ b/hazelcast/src/test/java/com/hazelcast/nio/tcp/FirewallingTcpIpConnectionManager.java
@@ -24,13 +24,12 @@ import com.hazelcast.nio.Connection;
 import com.hazelcast.nio.NodeIOService;
 
 import java.nio.channels.ServerSocketChannel;
-import java.util.Collections;
+import java.util.HashSet;
 import java.util.Set;
-import java.util.concurrent.ConcurrentHashMap;
 
 public class FirewallingTcpIpConnectionManager extends TcpIpConnectionManager {
 
-    private final Set<Address> blockedAddresses = Collections.newSetFromMap(new ConcurrentHashMap<Address, Boolean>());
+    private final Set<Address> blockedAddresses = new HashSet<Address>();
 
     public FirewallingTcpIpConnectionManager(
             LoggingService loggingService,
@@ -42,7 +41,7 @@ public class FirewallingTcpIpConnectionManager extends TcpIpConnectionManager {
     }
 
     @Override
-    public Connection getOrConnect(Address address, boolean silent) {
+    public synchronized Connection getOrConnect(Address address, boolean silent) {
         Connection connection = getConnection(address);
         if (connection != null) {
             return connection;
@@ -55,11 +54,11 @@ public class FirewallingTcpIpConnectionManager extends TcpIpConnectionManager {
         return super.getOrConnect(address, silent);
     }
 
-    public void block(Address address) {
+    public synchronized void block(Address address) {
         blockedAddresses.add(address);
     }
 
-    public void unblock(Address address) {
+    public synchronized void unblock(Address address) {
         blockedAddresses.remove(address);
         Connection connection = getConnection(address);
         if (connection instanceof DroppingConnection) {

